### ISSUE TITLE
xds: Envoy should support its built-in LB policies as extensions

### ISSUE THREAD

**Issue by markdroth** [envoyproxy/envoy#20634]
LB extensibility was originally described in #5598.  The infrastructure for this has been implemented, but there are a few missing pieces.  Here's the relevant history:
- A new field for LB policy extensibility was added to xDS in #7744.
- The extension mechanism was added to Envoy in #17400.
- The semantics of the xDS field were changed slightly in #18419.
- xDS extension protos were added for ROUND_ROBIN and RING_HASH in #19517 and #20423, but Envoy does not yet actually implement configuring those policies as extensions.

I think the remaining work here is:
- Add extension protos for all built-in LB policies.
- Implement actually configuring these policies via the extension protos.

Once this is done, we can consider whether we want to deprecate the existing enum for configuring built-in policies.

**Comment by daixiang0** [comment]
Compared with enum, I prefer LB policies.

**Comment by wbpcode** [comment]
Great. This can make the envoy lb policies to be easy extended.

**Comment by markdroth** [comment]
Looks like a config for the least_request policy is being added in #21210.

**Comment by markdroth** [comment]
It looks like Envoy currently requires that the `common_lb_config` field is not set if the new `load_balancing_policy` field is used:

https://cs.github.com/envoyproxy/envoy/blob/6c75133dac08e6db4988d398479dc02a5dd8f11c/source/common/upstream/upstream_impl.cc#L990

I think this needs to be changed such that if the `load_balancing_policy` field is used, Envoy simply ignores the `common_lb_config` field.  Otherwise, there's no way to cleanly switch from the old field to the new field as intended in https://github.com/envoyproxy/envoy/pull/18419.

**Comment by markdroth** [comment]
I don't think this is done yet.  We still need to actually implement the extensions, and figure out how to handle `CommonLbConfig`.

**Comment by wbpcode** [comment]
Seems like because of this. 🤣 

![image](https://user-images.githubusercontent.com/12389633/196468536-ed007215-078b-4259-8fd7-65a4564e765c.png)

**Comment by wbpcode** [comment]
Next we should make the subset lb as an extension.

**Comment by alyssawilk** [comment]
curious, is the plan here to port subset lb to be a thread aware load balancer, or extend TypedLoadBalancerFactory to be able to create non-thread-aware lbs? if the latter I'd be game for helping out!

**Comment by wbpcode** [comment]
@alyssawilk It would be great if you can help it. 😸 

I think both of these are out targes. The first one is a mid-term goal.
And the second one should be achieved partly for now. as we have created non-thread-aware lbs by the `TypedLoadBalancerFactory`. (Although I think there are some optimizations need to be landed.)

**Comment by alyssawilk** [comment]
have you? It looks like it only creates ThreadAwareLBs:
https://github.com/envoyproxy/envoy/blob/main/envoy/upstream/load_balancer.h#L250 ?

I was thinking of extending it to be able to create non-thread-aware so we can move out subset but wasn't sure if you were already on it, or had different plans :-)

**Comment by wbpcode** [comment]
cc @markdroth I think this is basically done.

Although I will do some more optimization to it, like docs, examples, code optimizations, API frozen, etc.

**Comment by markdroth** [comment]
Awesome!  Thank you so much for doing this!
